src/Cardano/Wallet/Kernel/Wallet.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE TypeOperators #-}
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- | Module provides internal implementation of
+-- | wallet api
+
+module Cardano.Wallet.Kernel.Wallet where
+
+import Prelude
+
+import Cardano.Wallet
+    ( Wallet )
+import Cardano.Wallet.AddressDiscovery
+    ( AddressPoolGap )
+import Cardano.Wallet.Mnemonic
+    ( Mnemonic )
+import Cardano.Wallet.Primitive
+    ( Coin )
+import Crypto.Encoding.BIP39
+    ( ValidMnemonicSentence )
+import Data.ByteArray
+    ( ScrubbedBytes )
+import Data.Text
+    ( Text )
+import Data.Time.Units
+    ( Microsecond )
+import Fmt
+    ( Buildable (..), fmt )
+
+
+-- | Errors
+data CreateWalletError = CreateWalletError
+
+instance Buildable CreateWalletError where
+    build CreateWalletError =
+        mempty
+        <> "CreateWalletError"
+
+instance Show CreateWalletError where
+    show = fmt . build
+
+data GetWalletError = GetWalletError
+
+instance Buildable GetWalletError where
+    build GetWalletError =
+        mempty
+        <> "GetWalletError"
+
+instance Show GetWalletError where
+    show = fmt . build
+
+
+-- | Types
+newtype CreateWallet = CreateWallet NewWallet
+
+data NewWallet =
+    forall mw . ValidMnemonicSentence mw => NewWallet
+    {
+      _mnemonicSentence
+        :: !(Mnemonic mw)
+    , _mnemonicSentencePassphrase
+        :: !(Maybe (Mnemonic mw))
+    , _name
+        :: !WalletName
+    , _passphrase
+        :: !WalletPassPhrase
+    , _addressPoolGap
+        :: !AddressPoolGap
+    }
+
+newtype WalletName = WalletName { _getWalletName :: Text }
+    deriving (Eq, Show)
+
+newtype WalletPassPhrase = WalletPassPhrase ScrubbedBytes
```

Comment:
Could we instead use `Passphrase "generation"` defined in `AddressDerivation`?

Reply:
definitely!